plugin/src/main/java/org/opensearch/sql/plugin/catalog/CatalogServiceImpl.java

```diff
@@ -0,0 +1,164 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.sql.catalog.CatalogService;
+import org.opensearch.sql.opensearch.security.SecurityAccess;
+import org.opensearch.sql.opensearch.setting.OpenSearchSettings;
+import org.opensearch.sql.plugin.model.CatalogMetadata;
+import org.opensearch.sql.plugin.model.ConnectorType;
+import org.opensearch.sql.storage.StorageEngine;
+
+/**
+ * This class manages catalogs and responsible for creating connectors to these catalogs.
+ */
+public class CatalogServiceImpl implements CatalogService {
+
+  private static final CatalogServiceImpl INSTANCE = new CatalogServiceImpl();
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  public static final String OPEN_SEARCH = "opensearch";
+
+  private Map<String, StorageEngine> storageEngineMap = new HashMap<>();
+
+  public static CatalogServiceImpl getInstance() {
+    return INSTANCE;
+  }
+
+  private CatalogServiceImpl() {
+  }
+
+  /**
+   * This function reads settings and loads connectors to the data stores.
+   * This will be invoked during start up and also when settings are updated.
+   *
+   * @param settings settings.
+   */
+  public void loadConnectors(Settings settings) {
+    doPrivileged(() -> {
+      Boolean isPPLEnabled = (Boolean) OpenSearchSettings.PPL_ENABLED_SETTING.get(settings);
+      if (isPPLEnabled) {
+        InputStream inputStream = CatalogSettings.CATALOG_CONFIG.get(settings);
+        if (inputStream != null) {
+          ObjectMapper objectMapper = new ObjectMapper();
+          objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+          try {
+            List<CatalogMetadata> catalogs =
+                objectMapper.readValue(inputStream, new TypeReference<>() {
+                });
+            LOG.info(catalogs.toString());
+            validateCatalogs(catalogs);
+            constructConnectors(catalogs);
+          } catch (IOException e) {
+            LOG.error("Catalog Configuration File uploaded is malformed. Verify and re-upload.");
+            throw new IllegalArgumentException(
+                "Malformed Catalog Configuration Json" + e.getMessage());
+          }
+        }
+      } else {
+        LOG.info("PPL federation is not enabled.");
+      }
+      return null;
+    });
+  }
+
+  @Override
+  public StorageEngine getStorageEngine(String catalog) {
+    if (catalog == null || !storageEngineMap.containsKey(catalog)) {
+      return storageEngineMap.get(OPEN_SEARCH);
+    }
+    return storageEngineMap.get(catalog);
+  }
+
+  @Override
+  public Set<String> getCatalogs() {
+    Set<String> catalogs = storageEngineMap.keySet();
+    catalogs.remove(OPEN_SEARCH);
+    return catalogs;
+  }
+
+  @Override
+  public void registerOpenSearchStorageEngine(StorageEngine storageEngine) {
```

Comment:
nit: registerStorageEngine?

Reply:
This is only for OpensearchStorageEngine.  May be I should change the name to opensearchStorageEngine. We want to put under key "opensearch". Rest of the storage engines should be initialized only during SQLPlugin initialization.